### DISCUSSION TITLE
Intellisense not working for Playwright, REST

### DISCUSSION THREAD

**Discussion by zhigulin-n** [codeceptjs/CodeceptJS#3553] [Q&A] [answered]
I write autotests on the Codeceptjs framework in TypeScript. After adding additional modules to the project, as well as after updating the repository with the command:

`npx codeceptjs def`

Intellisense is starting to work for me and I'm starting to see methods that I can use. For example, I see methods of such modules as: JSONResponse, DbHelper, ChaiWrapper.
I also installed Playwright and REST. But for these modules, intelisense does not work.

```
import { setHeadlessWhen, setCommonPlugins } from '@codeceptjs/configure';
// turn on headless mode when running with HEADLESS=true environment variable
// export HEADLESS=true && npx codeceptjs run
setHeadlessWhen(process.env.HEADLESS);
const env = require('dotenv').config({
path: '.env.test'
})

if (env.error && env.error.code === 'ENOENT') {
throw new Error('Please, recreate env.test file')
}

const base_url = process.env.URL
const currentBrowser = process.env.profile ?? 'chromium'; // 'firefox', 'webkit'

// enable all common plugins https://github.com/codeceptjs/configure#setcommonplugins
setCommonPlugins();

export const config: CodeceptJS.MainConfig = {
tests: './tests/\*\_test.ts',
output: './output',
helpers: {
ChaiWrapper: {
require: "codeceptjs-chai"
},
Playwright: {
url: base_url,
show: true,
browser: currentBrowser
},
REST: {
endpoint: base_url,
},
JSONResponse: {},
DbHelper: {
"require": "codeceptjs-dbhelper"
}
},
include: {
"I": "./common/custom_steps",
},
plugins: {
tryTo: {
enabled: true
}
},
name: 'test',
fullPromiseBased: false
}
```

package.json

```
{
"name": "codeceptjs-tests",
"version": "0.1.0",
"private": true,
"scripts": {
"codeceptjs": "codeceptjs run --steps",
"codeceptjs:headless": "HEADLESS=true codeceptjs run --steps",
"codeceptjs:ui": "codecept-ui --app",
"codeceptjs:demo": "codeceptjs run --steps -c node_modules/@codeceptjs/examples",
"codeceptjs:demo:headless": "HEADLESS=true codeceptjs run --steps -c node_modules/@codeceptjs/examples",
"codeceptjs:demo:ui": "codecept-ui --app  -c node_modules/@codeceptjs/examples"
},
"devDependencies": {
"@codeceptjs/configure": "^0.10.0",
"@codeceptjs/examples": "^1.2.1",
"@codeceptjs/ui": "^0.4.7",
"@types/node": "^18.11.12",
"codeceptjs": "^3.3.7",
"playwright": "^1.28.1",
"ts-node": "^10.9.1",
"typescript": "^4.9.4"
},
"dependencies": {
"codeceptjs-chai": "^2.3.3",
"codeceptjs-dbhelper": "^1.2.2",
"database-js-mssql": "^0.3.0",
"database-js-mysql": "^1.1.3",
"dateformat": "^5.0.3",
"dotenv": "^16.0.3",
"faker": "^6.6.6",
"object-hash": "^3.0.0",
"randexp": "^0.5.3",
"random-date-generator": "^1.0.2",
"random-mobile-numbers": "^1.0.1",
"underscore": "^1.13.6",
"unique-names-generator": "^4.7.1"
}
}
```

Reinstalling modules doesn't help.

**Comment by kobenguyent** [comment]
Did you run npx codeceptjs def once you added Playwright and REST?

**Comment by zhigulin-n** [comment]
> Did you run npx codeceptjs def once you added Playwright and REST?

Yes

**Reply by kobenguyent** [reply]
May you share the step.d.ts if that's possible?

**Reply by zhigulin-n** [reply]
sure

```
/// <reference types='codeceptjs' />
type steps_file = typeof import('./common/custom_steps');
type ChaiWrapper = import('codeceptjs-chai');
type DbHelper = import('codeceptjs-dbhelper');

declare namespace CodeceptJS {
  interface SupportObject { I: I, current: any }
  interface Methods extends ChaiWrapper, Playwright, REST, JSONResponse, DbHelper {}
  interface I extends ReturnType<steps_file>, WithTranslation<ChaiWrapper>, WithTranslation<JSONResponse>, WithTranslation<DbHelper> {}
  namespace Translation {
    interface Actions {}
  }
}
```

**Reply by myrepojuly** [reply]
I see an issue with the steps_file in the above steps.d.ts. 

`type steps_file = typeof import('./common/custom_steps');` Try replacing it with 

`type steps_file = typeof import('./steps_file.js');`

**Reply by myrepojuly** [reply]
I guess the problem is from this file which is included in the config above.
```
 include: {
"I": "./common/custom_steps",
},
```

**Reply by kobenguyent** [reply]
include: {
    I: './steps_file.ts'
  },

Perhaps the extension is needed.

**Reply by zhigulin-n** [reply]
It didn't help :(

**Reply by zhigulin-n** [reply]
> I see an issue with the steps_file in the above steps.d.ts.
> 
> `type steps_file = typeof import('./common/custom_steps');` Try replacing it with
> 
> `type steps_file = typeof import('./steps_file.js');`

that didn't help too

**Reply by zhigulin-n** [reply]
My steps_file is in the correct location and has the correct name. Methods from it are available to me.

**Reply by kobenguyent** [reply]
Well, I guess, it should be

type steps_file = typeof import('./common/custom_steps.ts'); or type steps_file = typeof import('./common/custom_steps.js'); depends on your created file.

**Reply by zhigulin-n** [reply]
> Well, I guess, it should be
> 
> type steps_file = typeof import('./common/custom_steps.ts'); or type steps_file = typeof import('./common/custom_steps.js'); depends on your created file.

Had no effect

**Reply by kobenguyent** [reply]
https://github.com/PeterNgTr/thanh-nguyen/tree/main/task2

May you try this demo project. Perhaps you can understand what went wrong with yours.

**Reply by zhigulin-n** [reply]
This is one of the solutions to my problem, I already found a couple of templates for the project.

But for now, I'm trying to figure it out specifically in my project.

**Reply by myrepojuly** [reply]
@zhigulin-n Just try restarting VSCode if you feel you have everything correct….

**Comment by zhigulin-n** [comment, marked as answer]
I found the solution. In **steps.d.ts** you need to add the transfer of Methods to WithTranslation

Instead of: 
`interface I extends ReturnType<steps_file>, WithTranslation<ChaiWrapper>, WithTranslation<JSONResponse>, WithTranslation<DbHelper> {}`

We are adding 
`interface I extends ReturnType<steps_file>, WithTranslation<Methods> {}`
